pkg/grafana/organization.go

```diff
@@ -0,0 +1,145 @@
+package grafana
+
+import (
+	"context"
+	stderrors "errors"
+
+	"github.com/pkg/errors"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/giantswarm/observability-operator/api/v1alpha1"
+)
+
+func (s *Service) SetupOrganization(ctx context.Context, grafanaOrganization *v1alpha1.GrafanaOrganization) error {
+	// Configure the organization in Grafana
+	if err := s.ConfigureOrganization(ctx, grafanaOrganization); err != nil {
+		return errors.WithStack(err)
+	}
+
+	var errs []error
+
+	// Update the datasources in the CR's status
+	if err := s.ConfigureDatasources(ctx, grafanaOrganization); err != nil {
+		errs = append(errs, err)
+	}
+
+	// Configure Grafana RBAC
+	if err := s.ConfigureGrafanaSSO(ctx); err != nil {
+		errs = append(errs, err)
+	}
+
+	if len(errs) > 0 {
+		return errors.WithStack(stderrors.Join(errs...))
+	}
+
+	return nil
+}
+
+func (s *Service) UnsetupOrganization(ctx context.Context, grafanaOrganization *v1alpha1.GrafanaOrganization) error {
+	logger := log.FromContext(ctx)
+
+	// Delete organization in Grafana if it exists
+	var organization = NewOrganization(grafanaOrganization)
+	if grafanaOrganization.Status.OrgID > 0 {
+		err := s.DeleteOrganization(ctx, organization)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+
+		grafanaOrganization.Status.OrgID = 0
+		if err = s.client.Status().Update(ctx, grafanaOrganization); err != nil {
```

Comment:
I'm not sure if having this here and not in the controller makes sense though. This is not business logic but a side effect that should be in the controller imo

Reply:
True, but that's used in 2 steps I am not sure how to extract this back into the controller.

Reply:
For the delete, I think it's easy, you just set it to 0 after the call to DeleteOrganization.

For the setup, I think we should either call status update only once, or setup in 2 steps. Maybe have a SetupOrganization that does create the org and configure sso and have another function named CreateDatasources that would also bé called in the controller? Then Moving the status update in the controller should en easy right?

Reply:
Alright, I moved the status update back into the controller. I did split the organization setup in 2 steps and called status update in the middle from the controller.